File: SimSim/Modules/Realm/Realm.m
//
//  Realm.m
//  SimSim
//
//  Created by Jesus Lopez on 25/04/2017.
//  Copyright © 2017 DaniilSmelov. All rights reserved.
//

#import "Realm.h"
#import <Cocoa/Cocoa.h>
#import "SimSim-Swift.h"

#define PATHS_REALM_FILES  [NSArray arrayWithObjects: @"Documents", @"Library/Caches", nil]
#define REALM_APP_NAME          @"Realm Browser"
#define REALM_APP_URL           @"http://itunes.apple.com/es/app/realm-browser/id1007457278"

@implementation Realm


- (void)generateRealmMenuForPath:(NSString *)aPath forMenu:(NSMenu *)menu withHotKey:(NSNumber *)hotkey icon:(NSImage *)icon {

    NSArray *realmFiles = [[self class] findRealmFiles:aPath];

    if ([realmFiles count] == 0) { return; }  // Skip if there are no realm files

    BOOL isRealmBrowserInstalled = [[self class] isRealmBrowserAvailable];
    NSMenuItem* realmMenuItem = nil;

    if (isRealmBrowserInstalled == false) { // There is at least one realm file but no realmbrowser installed
        realmMenuItem = [[NSMenuItem alloc] initWithTitle:@"Install Realm Browser" action:@selector(installRealmBrowser:) keyEquivalent:[hotkey stringValue]];
        [realmMenuItem setTarget:self];
        [realmMenuItem setRepresentedObject:REALM_APP_URL];
    }
    else if ([realmFiles count] == 1) { // There is exactly one realm file
        realmMenuItem = [[NSMenuItem alloc] initWithTitle:@"Realm" action:@selector(openRealmFile:) keyEquivalent:[hotkey stringValue]];
        [realmMenuItem setRepresentedObject:aPath];
        [realmMenuItem setImage:icon];
        [realmMenuItem setTarget:self];
        [realmMenuItem setRepresentedObject:[[realmFiles firstObject] fullPath]];
    }
    else {  // There is more than one realm file
        realmMenuItem = [[NSMenuItem alloc] initWithTitle:@"Realm" action:nil keyEquivalent:[hotkey stringValue]];
        [realmMenuItem setRepresentedObject:aPath];
        [realmMenuItem setImage:icon];


        NSMenu* menuRealm = [[NSMenu alloc] initWithTitle:@"Realm Browser"];
        [menuRealm setAutoenablesItems:NO];

        for (RealmFile *realmFile in realmFiles) {
            NSMenuItem* menuItem = [[NSMenuItem alloc] initWithTitle:[realmFile fileName] action:@selector(openRealmFile:) keyEquivalent:@""];
            [menuItem setTarget:self];
            [menuItem setRepresentedObject:[realmFile fullPath]];
            [menuItem setEnabled:isRealmBrowserInstalled];;
            [menuRealm addItem:menuItem];
        }
        
        [menu setSubmenu:menuRealm forItem:realmMenuItem];
    }
    
    [menu addItem:realmMenuItem];
}


- (void)openRealmFile:(id)sender {
    [[self class] openInRealmBrowser:[sender representedObject]];
}

- (void)installRealmBrowser:(id)sender {
    [self openUrl:[sender representedObject]];
}

- (void)openUrl:(NSString *)aUrl {
    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:aUrl]];
}

#pragma mark - Statics methods

+ (NSArray *)findRealmFiles:(NSString *)aPath {
    NSMutableArray *files = [NSMutableArray new];

    for (NSString *realmPath in PATHS_REALM_FILES) {
        NSString *folderPath = [NSString stringWithFormat:@"%@/%@", aPath, realmPath];
        NSArray *allFilesOfFolder = [Tools getSortedFilesFromFolder: folderPath];

        for (NSDictionary *file in allFilesOfFolder)
        {
            NSString *fileName = [Tools getNameFrom:file];

            if ([[fileName pathExtension] isEqualToString: @"realm"] == false) { continue; }    // Skip if not a realm file

            RealmFile *realmFile = [[RealmFile alloc] init];
            realmFile.fileName = fileName;
            realmFile.path = folderPath;

            [files addObject: realmFile];
        }
    }

    return files;
}

+ (NSArray *)removeNonRealmObject:(NSArray *)allFiles {
    NSPredicate *endsRealm = [NSPredicate predicateWithFormat:@"self ENDSWITH '.realm'"];
    return [allFiles filteredArrayUsingPredicate:endsRealm];
}

#pragma mark - Public Statics methods

+ (BOOL)isRealmBrowserAvailable {
    NSFileManager* fileManager = [NSFileManager defaultManager];
    return [fileManager fileExistsAtPath:[[self class] applicationPath]];
}

+ (NSString *)applicationPath {
    return [NSString stringWithFormat:@"/Applications/%@.app", REALM_APP_NAME];
}

+ (BOOL)isRealmAvailableForPath:(NSString *)aPath {
    NSArray *realmFiles = [[self class] findRealmFiles:aPath];
    return (realmFiles != nil && [realmFiles count] > 0);
}

+ (void)openInRealmBrowser:(NSString *)aPath {
    [[NSWorkspace sharedWorkspace] openFile:aPath withApplication:REALM_APP_NAME];
}

@end


@implementation RealmFile
- (NSString *)fullPath {
    return [NSString stringWithFormat:@"%@/%@", [self path], [self fileName]];
}
@end

File: SimSim/Realm.swift
import Foundation
import Cocoa

//============================================================================
class RealmFile: NSObject
{
    var fileName = ""
    var path = ""
    {
        didSet
        {
            // Make sure that path always ends with a slash
            if !path.hasSuffix("/")
            {
                path += "/"
            }
        }
    }

    //----------------------------------------------------------------------------
    func fullPath() -> String?
    {
        return path + fileName
    }
}

//============================================================================
class Realm: NSObject
{
    //----------------------------------------------------------------------------
    class func insertInstallerItem(for menu: NSMenu, withHotKey hotkey: NSNumber)
    {
        // There is at least one realm file but no realmbrowser installed
        let item = NSMenuItem(title: "Install Realm Studio", action: #selector(Realm.installRealmBrowser(_:)), keyEquivalent: "\(hotkey)")
        
        item.target = self
        item.representedObject = Constants.Realm.appUrl

        menu.addItem(item)
    }
    
    //----------------------------------------------------------------------------
    class func insertRealmFiles(files: NSArray, forPath aPath: String, for menu: NSMenu,
                                withHotKey hotkey: NSNumber, icon: NSImage)
    {
        var item: NSMenuItem
        
        if files.count == 1
        {
            // There is exactly one realm file
            item = NSMenuItem(title: "Realm Studio", action: #selector(Realm.openRealmFile(_:)), keyEquivalent: "\(hotkey)")
            item.representedObject = aPath
            item.image = icon
            item.target = self
            
            item.representedObject = (files.firstObject as! RealmFile).fullPath()
        }
        else
        {
            // There is more than one realm file
            item = NSMenuItem(title: "Realm Studio", action: nil, keyEquivalent: "\(hotkey)")
            item.representedObject = aPath
            item.image = icon
            
            let menuRealm = NSMenu(title: "Realm Studio")
            menuRealm.autoenablesItems = false
            
            let browserInstalled = isRealmBrowserAvailable()
            for file in files
            {
                let realmFile = file as! RealmFile
                let menuItem = NSMenuItem(title: realmFile.fileName as String, action: #selector(Realm.openRealmFile(_:)), keyEquivalent: "")
                menuItem.target = self
                menuItem.representedObject = realmFile.fullPath()
                menuItem.isEnabled = browserInstalled
                menuRealm.addItem(menuItem)
            }
            menu.setSubmenu(menuRealm, for: item)
        }
        menu.addItem(item)
    }
    
    //----------------------------------------------------------------------------
    class func generateRealmMenu(forPath aPath: String, for menu: NSMenu, withHotKey hotkey: NSNumber, icon: NSImage)
    {
        let files = findRealmFiles(aPath)!
        
        if files.count == 0
        {
            return
        }
        
        // Skip if there are no realm files
        if !isRealmBrowserAvailable()
        {
            insertInstallerItem(for: menu, withHotKey: hotkey)
        }
        else
        {
            insertRealmFiles(files: files, forPath: aPath, for: menu, withHotKey: hotkey, icon: icon)
        }

    }

    //----------------------------------------------------------------------------
    @objc
    class func openRealmFile(_ sender: NSMenuItem)
    {
        open(inRealmBrowser: sender.representedObject as! String)
    }
    
    //----------------------------------------------------------------------------
    @objc
    class func installRealmBrowser(_ sender: NSMenuItem)
    {
        openUrl(sender.representedObject as! String)
    }
    
    //----------------------------------------------------------------------------
    class func openUrl(_ aUrl: String)
    {
        if let anUrl = URL(string: aUrl)
        {
            NSWorkspace.shared.open(anUrl)
        }
    }

    //----------------------------------------------------------------------------
    class func findRealmFiles(_ aPath: String) -> NSArray?
    {
        let files = NSMutableArray()
        for realmPath: String? in Constants.Realm.dbPaths
        {
            guard realmPath != nil else
            {
                break
            }
            
            let folderPath = aPath + realmPath!
            
            let allFilesOfFolder = Tools.getSortedFiles(fromFolder: folderPath)
            
            for file in allFilesOfFolder
            {
                let fileName = Tools.getName(from: file) as NSString
                
                if (fileName.pathExtension == "realm")
                {
                    let realmFile = RealmFile()
                    realmFile.fileName = fileName as String
                    realmFile.path = folderPath
                    files.add(realmFile)
                }
            }
        }
        return files
    }

    //----------------------------------------------------------------------------
    class func removeNonRealmObject(_ allFiles: NSArray) -> NSArray
    {
        let endsRealm = NSPredicate(format: "self ENDSWITH '.realm'")
        return allFiles.filtered(using: endsRealm) as NSArray
    }

    //----------------------------------------------------------------------------
    class func isRealmBrowserAvailable() -> Bool
    {
        let fileManager = FileManager.default
        return fileManager.fileExists(atPath: applicationPath())
    }
    
    //----------------------------------------------------------------------------
    class func applicationPath() -> String
    {
        return Constants.Paths.realmApp
    }
    
    //----------------------------------------------------------------------------
    class func isRealmAvailable(forPath aPath: String) -> Bool
    {
        let realmFiles = findRealmFiles(aPath)
        return realmFiles != nil && realmFiles!.count > 0
    }
    
    //----------------------------------------------------------------------------
    class func open(inRealmBrowser aPath: String)
    {
        NSWorkspace.shared.openFile(aPath, withApplication: Constants.Realm.appName)
    }
}
